vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,799 @@
+                defaultTraverse(expr)
+            
+            override this.VisitInheritSynMemberDefn (_, _, _, _, range) = Some range
+
+            override this.VisitTypeAbbrev( _, range) = Some range
+
+            override this.VisitLetOrUse(binding, range) = Some range
+
+            override this.VisitComponentInfo componentInfo = Some componentInfo.Range
+        })
+
+    let formatMap = lazy typeMap.Value.ClassificationFormatMapService.GetClassificationFormatMap "tooltip"
+    //let visibleAdornments = ConcurrentDictionary()
+
+    let mutable lastResults = Dictionary<string, ITrackingSpan * CodeLens>()
+    let mutable firstTimeChecked = false
+    let mutable bufferChangedCts = new CancellationTokenSource()
+    let uiContext = SynchronizationContext.Current
+
+    let FSharpRangeToSpan (bufferSnapshot:ITextSnapshot) (range:range) =
+        try
+            let startLine, endLine = 
+                bufferSnapshot.GetLineFromLineNumber(range.StartLine - 1),
+                bufferSnapshot.GetLineFromLineNumber(range.EndLine - 1)
+            let startPosition = startLine.Start.Add range.StartColumn
+            let endPosition = endLine.Start.Add range.EndColumn
+            Span(startPosition.Position, endPosition.Position - startPosition.Position) |> Some
+        with e -> 
+            logErrorf "Error: %A" e
+            None
+
+    let layoutTagToFormatting (layoutTag: LayoutTag) =
+        layoutTag
+        |> RoslynHelpers.roslynTag
+        |> ClassificationTags.GetClassificationTypeName
+        |> typeMap.Value.GetClassificationType
+        |> formatMap.Value.GetTextProperties   
+
+    let createTextBox (lens:CodeLens) =
+        asyncMaybe {
+            let! taggedText, navigation = lens.TaggedText
+            do! Async.SwitchToContext uiContext |> liftAsync
+            let textBox = new TextBlock(Width = 100., Background = Brushes.Transparent, Opacity = 0.5, TextTrimming = TextTrimming.WordEllipsis)
+            DependencyObjectExtensions.SetDefaultTextProperties(textBox, formatMap.Value)
+            for text in taggedText do
+                let run = Documents.Run text.Text
+                DependencyObjectExtensions.SetTextProperties (run, layoutTagToFormatting text.Tag)
+                let inl =
+                    match text with
+                    | :? Layout.NavigableTaggedText as nav when navigation.IsTargetValid nav.Range ->
+                        let h = Documents.Hyperlink(run, ToolTip = nav.Range.FileName)
+                        h.Click.Add (fun _ -> navigation.NavigateTo nav.Range)
+                        h :> Documents.Inline
+                    | _ -> run :> _
+                textBox.Inlines.Add inl
+            textBox.Opacity <- 0.5
+            lens.Computed <- true
+            lens.UiElement <- textBox
+        } |> Async.Ignore
+
+    let executeCodeLenseAsync () =  
+        asyncMaybe {
+            do! Async.Sleep 800 |> liftAsync
```

Comment:
Magic sleeps again

Reply:
The magic sleeps are coupled to the cancellation tokens, they are needed to delay execution. We don't want this code to be executed after every single char we type.